lib/topological_inventory/providers/common/operations/source.rb

```diff
@@ -57,32 +57,57 @@ def params_missing?
         end
 
         def checked_recently?
-          return false if endpoint.nil?
+          checked_recently = if endpoint.present?
+                               endpoint.last_checked_at.present? && endpoint.last_checked_at >= LAST_CHECKED_AT_THRESHOLD.ago
+                             elsif application.present?
+                               application.last_checked_at.present? && application.last_checked_at >= LAST_CHECKED_AT_THRESHOLD.ago
+                             end
 
-          checked_recently = endpoint.last_checked_at.present? && endpoint.last_checked_at >= LAST_CHECKED_AT_THRESHOLD.ago
-          logger.availability_check("Skipping, last check at #{endpoint.last_checked_at} [Source ID: #{source_id}] ") if checked_recently
+          logger.availability_check("Skipping, last check at #{endpoint.last_checked_at || application.last_checked_at} [Source ID: #{source_id}] ") if checked_recently
 
           checked_recently
         end
 
         def connection_status
-          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:endpoint_not_found]] unless endpoint
-          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+          # we need either an endpoint or application to check the source.
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:endpoint_or_application_not_found]] unless endpoint || application
 
           check_time
+          if endpoint
+            endpoint_connection_check
+          elsif application
+            application_connection_check
+          end
+        end
+
+        def endpoint_connection_check
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+
+          # call down into the operations pod implementation of `Source#connection_check`
           connection_check
         end
 
+        def application_connection_check
+          case application.availability
```

Comment:
who is setting this value?
Is it by https://github.com/RedHatInsights/sources-api/blob/master/app/controllers/api/v1/sources_controller.rb#L56 call?

Reply:
btw I don't see `application.availability` in Sources API

Reply:
Yeah I typo'd that, it's updated now!

Reply:
also I talked Doug in cost, they're updating it from their end every time an availability request comes in.